### PR TITLE
feat: add user activity search component

### PR DESCRIPTION
**Pull request by BLasan**
## Description
Added user activity search component in Home page. Currently it supports for the activities mentioned in home.component.ts. Can add all the activities with relevant routings after integrating all branches.

## Related issues and discussion
#710 

## Screenshots, if any
webapp1

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

**Comment by BLasan** [comment]
Failed due to Travis build time out

**Comment by BLasan** [comment]
@abhaychawla  Please review this

File: src/app/home/home.component.ts [hunk 1]
```diff
@@ -16,18 +17,51 @@ export class HomeComponent implements OnInit {
...
+    { activity: 'groups', path: '/groups'},
+    { activity: 'centers', path: '/centers'},
+    { activity: 'accounting', path: '/accounting'},
+    { activity: 'users', path: '/users'},
+    { activity: 'organization', path: '/organization'},
+    { activity: 'system', path: '/system'},
+    { activity: 'templates', path: '/templates'},
+    { activity: 'self-service users', path: '/self-service/users'},
+    { activity: 'self-service app-configuration', path: '/self-service/app-configuration'},
+    { activity: 'task management', path: '/self-service/task-management'}
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:38]: @BLasan add all routes as mentioned in ExpertSearchController.js file in community app.

Comment (BLasan) [reply src/app/home/home.component.ts:38]: Sure. I'll update and send the PR soon.

Comment (BLasan) [reply src/app/home/home.component.ts:38]: > @BLasan add all routes as mentioned in ExpertSearchController.js file in community app.

Is it okay to use all the paths mentioned in community-app?

**Comment by totalnewb1** [comment]
yes even if they don't work right now. I will check if some change is
required.

File: src/app/home/home.component.ts (hunk 1 above)
Comment (karantakalkar) [reply src/app/home/home.component.ts:38]: yes even if they don't work right now, I will check if some route change is required.

Comment (BLasan) [reply src/app/home/home.component.ts]: Updated the activities mentioned in ExpertSearchController.js. Please review

File: src/app/home/home.component.scss [hunk 2]
```diff
@@ -6,3 +6,30 @@
...
+  width: 100%;
+  height: auto;
+  background:white;
+  top: 105px;
+  left: 0;
+  transition: 0.2s all;
+  box-shadow: 0 0 10px 0 rgba(0,0,0,0.5);
+  border-radius:3px;
+  display: block
+}
```
Comment (karantakalkar) [inline src/app/home/home.component.scss:22]: use mat select.
```suggestion
}
```

File: src/app/home/home.component.html [hunk 3]
```diff
@@ -17,6 +17,27 @@
...
+      <mat-card-content>
+
+        <div fxLayout="row" fxLayoutGap="20px">
+          <mat-form-field fxFlex>
+            <mat-label> Search Activities </mat-label>
+            <input matInput formControlName="searchText">
+          </mat-form-field>
+        </div>
+
+        <div class="search-result" style="display:none" id="search_dropdown">
```
Comment (karantakalkar) [inline src/app/home/home.component.html:31]: use mat select instead of this.

File: src/app/home/home.component.html [hunk 4]
```diff
@@ -17,6 +17,27 @@
       </mat-card-title>
     </mat-card-header>
 
+    <form [formGroup]="activityForm">
```
Comment (karantakalkar) [inline src/app/home/home.component.html:20]: no need for form group a single formcontrol will do

File: src/app/home/home.component.scss [hunk 5]
```diff
@@ -6,3 +6,30 @@
...
+
+a{
+  text-decoration: none;
+  color: black
+}
+
+mat-list-item:hover{
+  cursor: pointer;
+}
```
Comment (karantakalkar) [inline src/app/home/home.component.scss:35]: remove any unused scss, add new line at EOF

File: src/app/home/home.component.ts [hunk 6]
```diff
@@ -16,18 +17,99 @@ export class HomeComponent implements OnInit {
...
   ngOnInit() {
     const credentials = this.authenticationService.getCredentials();
     this.username = credentials.username;
+    this.setSearchForm();
+  }
+
+  /**
+   * Set Search Form.
+   */
+  setSearchForm() {
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:109]: use single form control

File: src/app/pipes/search-activity.pipe.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { Pipe, PipeTransform } from '@angular/core';
```
Comment (karantakalkar) [inline src/app/pipes/search-activity.pipe.ts:1]: delete this pipe

File: src/app/shared/shared.module.ts [hunk 8]
```diff
@@ -46,7 +47,8 @@ import { MaterialModule } from './material.module';
     IconsModule,
     MaterialModule,
     FormsModule,
-    ReactiveFormsModule
+    ReactiveFormsModule,
```
Comment (karantakalkar) [inline src/app/shared/shared.module.ts:50]: remove this import

File: src/app/home/home.component.ts [hunk 9]
```diff
@@ -16,18 +17,99 @@ export class HomeComponent implements OnInit {
 
   /** Username of authenticated user. */
   username: string;
+  /** Check if searched. */
+  isSeearched = false;
```
Comment (abhaychawla) [inline src/app/home/home.component.ts:21]: This is not used anywhere.

File: src/app/shared/shared.module.ts (hunk 8 above)
Comment (BLasan) [reply src/app/shared/shared.module.ts:50]: When removing ReactiveFormsModule it's going to raise some issues in some components.

Comment (karantakalkar) [reply src/app/shared/shared.module.ts:50]: oh sorry remove the comma

File: src/app/shared/shared.module.ts [hunk 10]
```diff
@@ -15,6 +15,7 @@ import { ThemePickerComponent } from './theme-picker/theme-picker.component';
 /** Custom Modules */
 import { IconsModule } from './icons.module';
 import { MaterialModule } from './material.module';
+import { PipesModule } from 'app/pipes/pipes.module';
```
Comment (karantakalkar) [inline src/app/shared/shared.module.ts:18]: remove this

File: src/app/shared/shared.module.ts [hunk 11]
```diff
@@ -46,7 +47,8 @@ import { MaterialModule } from './material.module';
     IconsModule,
     MaterialModule,
     FormsModule,
-    ReactiveFormsModule
+    ReactiveFormsModule,
+    PipesModule
```
Comment (karantakalkar) [inline src/app/shared/shared.module.ts:51]: pipes module is to be imported seperately from shared, also here its not needed

File: src/app/home/home.module.ts [hunk 12]
```diff
@@ -8,6 +8,7 @@ import { HomeRoutingModule } from './home-routing.module';
 /** Custom Components */
 import { HomeComponent } from './home.component';
 import { DashboardComponent } from './dashboard/dashboard.component';
+import { ReactiveFormsModule } from '@angular/forms';
```
Comment (karantakalkar) [inline src/app/home/home.module.ts:11]: this comes from shared remove please

File: src/app/shared/shared.module.ts (hunk 8 above)
Comment (BLasan) [reply src/app/shared/shared.module.ts:50]: There's a pipe module imported after that

**Comment by BLasan** [comment]
@karantakalkar  Please review this

**Comment by karantakalkar** [comment]
@BLasan you can focus on this one till i review edit loan charge, check for corresponding routes in the web app for the ones in community app, for those whose components have not been developed leave blank route

**Comment by BLasan** [comment]
> @BLasan you can focus on this one till i review edit loan provision, check for corresponding routes in the web app for the ones in community app, for those whose components have not been developed leave blank route

Ok. I'll work on this

**Comment by BLasan** [comment]
@karantakalkar Refactored. Please review

File: src/app/home/home.component.html [hunk 13]
```diff
@@ -17,6 +17,23 @@
       </mat-card-title>
     </mat-card-header>
 
+    <mat-card-content>
+        
+      <div fxLayout="column">
+
+        <mat-form-field>
```
Comment (karantakalkar) [inline src/app/home/home.component.html:24]: refer to accounting section, make autocomplete HTML as well as code similar to auto completes being used in accounting section (for eg: gl account autocomplete)

File: src/app/home/home.component.ts [hunk 14]
```diff
@@ -16,18 +19,95 @@ export class HomeComponent implements OnInit {
...
+    { activity: 'enter collection sheet', path: ''},
+    { activity: 'product mix', path: '/products/products-mix'},
+    { activity: 'add product mix', path: ''},
+    { activity: 'bulk loan reassignment', path: '/accounting/closing-entries/create'},
+    { activity: 'audit', path: '/products/audit-trails'},
+    { activity: 'create accounting closure', path: '/createclosure'},
+    { activity: 'navigation', path: '/navigation'},
+    { activity: 'system users', path: '/aself-service/users'},
+    { activity: '', path: 'home'}
+  ];
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:88]: @BLasan shift this to a new file activities.ts, import it here

File: src/app/home/home.component.ts [hunk 15]
```diff
@@ -16,18 +19,95 @@ export class HomeComponent implements OnInit {
...
+    this.filteredActivities = this.searchText.valueChanges.pipe(startWith(''), map(value => value ? this.filterActivity(value) : this.allActivities.slice()));
   }
 
+  /**
+   * Filter Activity.
+   */
+  filterActivity(value: string) {
+    const filterValue = value.toLowerCase();
+    return this.allActivities.filter(activity => activity.activity.toLowerCase().indexOf(filterValue) === 0);
+  }
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:112]: check autocomplete code in accounting section for examples, make it exactly similar.

File: src/app/home/home.component.ts [hunk 16]
```diff
@@ -16,18 +19,95 @@ export class HomeComponent implements OnInit {
...
+    { activity: 'add member', path: '/clients/:clientId/family-members/add'},
+    { activity: 'create charge', path: ''},
+    { activity: 'enter collection sheet', path: ''},
+    { activity: 'product mix', path: '/products/products-mix'},
+    { activity: 'add product mix', path: ''},
+    { activity: 'bulk loan reassignment', path: '/accounting/closing-entries/create'},
+    { activity: 'audit', path: '/products/audit-trails'},
+    { activity: 'create accounting closure', path: '/createclosure'},
+    { activity: 'navigation', path: '/navigation'},
+    { activity: 'system users', path: '/aself-service/users'},
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:86]: check this route

File: src/app/home/home.component.ts [hunk 17]
```diff
@@ -16,18 +19,95 @@ export class HomeComponent implements OnInit {
...
+    { activity: 'account closure', path: '/accounting/closing-entries'},
+    { activity: 'accounting rules', path: '/accounting/accounting-rules'},
+    { activity: 'add accounting rule', path: '/accounting/accounting-rules/create'},
+    { activity: 'data tables', path: '/system/data-tables'},
+    { activity: 'create data table', path: '/system/data-tables/create'},
+    { activity: 'add code', path: '/system/codes/create'},
+    { activity: 'jobs', path: '/system/scheduler-jobs'},
+    { activity: 'codes', path: '/system/codes'},
+    { activity: 'reports', path: '/reports'},
+    { activity: 'create report', path: '/'},
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:74]:
```suggestion
    { activity: 'create report', path: ''},
```

File: src/app/home/home.component.html [hunk 18]
```diff
@@ -17,6 +17,23 @@
       </mat-card-title>
```
Comment (karantakalkar) [inline src/app/home/home.component.html:17]: align title with start of search field

File: src/app/home/home.component.ts [hunk 19]
```diff
@@ -16,18 +19,95 @@ export class HomeComponent implements OnInit {
...
+    { activity: 'create group', path: ''},
+    { activity: 'create center', path: '/centers/create'},
+    { activity: 'configuration', path: '/system/global-configurations'},
+    { activity: 'tasks', path: '/self-service/task-management'},
+    { activity: 'create template', path: '/self-service/task-management'},
+    { activity: 'create loan product', path: '/products/loan-products/create'},
+    { activity: 'create saving product', path: '/products/saving-products/create'},
+    { activity: 'roles', path: '/system/roles-and-permissions'},
+    { activity: 'add role', path: '/system/roles-and-permissions/add'},
+    { activity: 'configure maker checker tasks', path: '/system/global-configurations'},
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:50]: not implemented yet
```suggestion
    { activity: 'configure maker checker tasks', path: ''},
```

File: src/app/home/home.component.ts [hunk 20]
```diff
@@ -16,18 +19,95 @@ export class HomeComponent implements OnInit {
...
+    { activity: 'loan products', path: 'products/loan-products'},
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:51]: not working

File: src/app/home/home.component.ts [hunk 21]
```diff
@@ -16,18 +19,95 @@ export class HomeComponent implements OnInit {
...
+    { activity: 'charges', path: 'products/charges'},
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:52]: not working

File: src/app/home/home.component.ts [hunk 22]
```diff
@@ -16,18 +19,95 @@ export class HomeComponent implements OnInit {
...
+    { activity: 'saving products', path: 'products/saving-products'},
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:53]: not working

File: src/app/home/home.component.ts [hunk 23]
```diff
@@ -16,18 +19,95 @@ export class HomeComponent implements OnInit {
...
+    { activity: 'offices', path: 'organization/offices'},
+    { activity: 'create office', path: 'organization/offices/create'},
+    { activity: 'currency configurations', path: 'organization/currencies'},
+    { activity: 'user settings', path: ''},
+    { activity: 'employees', path: 'organization/employees'},
+    { activity: 'create employee', path: 'organization/employees/create'},
```
Comment (karantakalkar) [inline src/app/home/home.component.ts:59]: you have forgotten to add '/' at start of route

File: src/app/home/home.component.html (hunk 13 above)
Comment (BLasan) [reply src/app/home/home.component.html:24]: I have used mat-autocomplete. What's the error in here?

Comment (karantakalkar) [reply src/app/home/home.component.html:24]: There are multiple ways of implementing autocomplete, make it consistent with rest of app.

Comment (BLasan) [reply src/app/home/home.component.html:24]: Are there any autocomplete components similar to this? I found only filtering techniques which were used to filter the mat-table.

File: src/app/home/home.component.ts (hunk 23 above)
Comment (BLasan) [reply src/app/home/home.component.ts:59]: yes. I'll fix it. But can you guide me to implement the mat autocomplete as in the app? I did not find any similar components to this other than filtering used for mat tables

File: src/app/home/home.component.html (hunk 13 above)
Comment (karantakalkar) [reply src/app/home/home.component.html:24]: use your text editor to search for gl account autocomplete in code base, or just search for autocomplete, which code editor do you use?

Comment (BLasan) [reply src/app/home/home.component.html:24]: VS Code

Comment (BLasan) [reply src/app/home/home.component.html:24]: > use your text editor to search for gl account autocomplete in code base, or just search for autocomplete, which code editor do you use?

I searched this. But no result found

Comment (BLasan) [reply src/app/home/home.component.html:24]: Found in closing-entries. I'll fix. Thanks :)

Comment (BLasan) [reply src/app/home/home.component.html:23]: Done. Please review

File: src/app/home/home.component.ts (hunk 14 above)
Comment (BLasan) [reply src/app/home/home.component.ts]: Done

File: src/app/home/home.component.ts (hunk 15 above)
Comment (BLasan) [reply src/app/home/home.component.ts:69]: Done

File: src/app/home/home.component.ts (hunk 16 above)
Comment (BLasan) [reply src/app/home/home.component.ts]: Fixed

File: src/app/home/home.component.html (hunk 18 above)
Comment (BLasan) [reply src/app/home/home.component.html:17]: Done

**Comment by BLasan** [comment]
@karantakalkar Please review

**Comment by karantakalkar** [comment]
Looks good to me, can you make autocomplete panel visible only when user types something, ie; it should not open just by clicking the search field.
@muskankhedia please review(specially rxjs part).

**Comment by BLasan** [comment]
> Looks good to me, can you make autocomplete panel visible only when user types something, ie; it should not open just by clicking the search field.
> @muskankhedia please review(specially rxjs part).

Ok. I'll fix it

**Comment by BLasan** [comment]
> Looks good to me, can you make autocomplete panel visible only when user types something, ie; it should not open just by clicking the search field.
> @muskankhedia please review(specially rxjs part).

Done. Please review

**Comment by muskankhedia** [comment]
Some more routes need to be added, can be done after the GSoC period as most of the routes would be implemented till then.

**Comment by BLasan** [comment]
> @BLasan Good Work!

Thanks

**Comment by BLasan** [comment]
> Some more routes need to be added, can be done after the GSoC period as most of the routes would be implemented till then.

Ok :)

**Pull request merged**